### ISSUE TITLE
Web interface always predicts digit 8

### ISSUE THREAD

**Issue by maydaythecoder** [maydaythecoder/numbers#1]
## **Description**

The web UI consistently outputs the digit **8** with very high confidence, no matter what digit is drawn on the canvas. Even obvious inputs (1, 2, 3, 7, etc.) are predicted as **8**.

---

## **Steps to Reproduce**

1. Start the web app:

   ```
   python run.py  →  Select option 1
   ```
2. Open:
   **[http://localhost:5000/](http://localhost:5000)**
3. Draw any digit on the canvas.
4. Click **Predict**.
5. Observe the prediction result.

---

## **Expected Behavior**

The model should correctly classify the drawn digit after applying proper preprocessing.

## **Actual Behavior**

* Output is **always “8”**
* Confidence is consistently high
* Happens for all digits

---

## **Environment**

* Python **3.13**
* Flask (**debug mode**)
* TensorFlow/Keras model: **handwritten_digits.model.keras**

---

## **Relevant Code**

* Prediction route: **web_app.py:21–67**
* Image preprocessing: **web_app.py:38–51**
* Canvas capture code: **templates/index.html:243–254**

---

## **Initial Diagnosis**

The model itself works correctly (confirmed via `recognition.py`).
The issue is isolated to the **web preprocessing pipeline**.

### **Likely Causes**

* **Preprocessing mismatch** between canvas data and model training data
* **Color inversion error**: MNIST expects **white digit on black background**
* **Normalization producing uniform values**
* **Canvas → Base64 PNG conversion differences**
* Image may be:

  * Upscaled incorrectly
  * Not converted to grayscale properly
  * Thresholded incorrectly
  * Missing batch dimension
  * Extremely low-contrast after preprocessing
